internal/wal_verify_handler_test.go

```diff
@@ -0,0 +1,540 @@
+		"000000050000000000000001",
+		"000000050000000000000002",
+		"000000050000000000000003",
+		"000000050000000000000004",
+	}
+	currentSegmentName := "000000050000000000000005"
+	currentSegment, _ := internal.NewWalSegmentDescription(currentSegmentName)
+
+	expectedIntegrityCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusOk,
+		Details: internal.IntegrityCheckDetails{
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000001",
+				EndSegment:    "000000050000000000000004",
+				SegmentsCount: 4,
+				Status:        internal.Found,
+			},
+		},
+	}
+
+	expectedTimelineCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusOk,
+		Details: internal.TimelineCheckDetails{
+			CurrentTimelineId:        currentSegment.Timeline,
+			HighestStorageTimelineId: currentSegment.Timeline,
+		},
+	}
+
+	testWalVerify(t, WalVerifyTestSetup{
+		expectedIntegrityCheck: expectedIntegrityCheck,
+		expectedTimelineCheck:  expectedTimelineCheck,
+		currentWalSegment:      currentSegment,
+		storageSegments:        storageSegments,
+		storageFiles:           make(map[string]*bytes.Buffer),
+	})
+}
+
+// check that wal-verify correctly marks delayed segments
+func TestWalVerify_SingleTimeline_SomeDelayed(t *testing.T) {
+	storageSegments := []string{
+		"000000050000000000000001",
+		"000000050000000000000002",
+		"000000050000000000000003",
+		"000000050000000000000004",
+	}
+
+	currentSegmentName := "000000050000000000000019"
+	currentSegment, _ := internal.NewWalSegmentDescription(currentSegmentName)
+
+	expectedIntegrityCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusWarning,
+		Details: internal.IntegrityCheckDetails{
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000001",
+				EndSegment:    "000000050000000000000004",
+				SegmentsCount: 4,
+				Status:        internal.Found,
+			},
+			{
+				TimelineId:    5,
+				StartSegment:  "000000050000000000000005",
+				EndSegment:    "000000050000000000000018",
+				SegmentsCount: 20,
+				Status:        internal.ProbablyDelayed,
+			},
+		},
+	}
+
+	expectedTimelineCheck := internal.WalVerifyCheckResult{
+		Status: internal.StatusOk,
+		Details: internal.TimelineCheckDetails{
+			CurrentTimelineId:        currentSegment.Timeline,
+			HighestStorageTimelineId: currentSegment.Timeline,
+		},
+	}
+
+	testWalVerify(t, WalVerifyTestSetup{
+		expectedIntegrityCheck: expectedIntegrityCheck,
+		expectedTimelineCheck:  expectedTimelineCheck,
+		currentWalSegment:      currentSegment,
+		storageSegments:        storageSegments,
+		storageFiles:           make(map[string]*bytes.Buffer),
+	})
+}
+
+// check that wal-verify correctly follows timeline switches
+func TestWalVerify_TwoTimelines_Ok(t *testing.T) {
+	storageSegments := []string{
+		"000000050000000000000001",
+		"000000050000000000000002",
+		"000000050000000000000003",
+		"000000050000000000000004",
+		"000000050000000000000005", // should not get into output
```

Comment:
@x4m take a look. Is it okay if we have some more WAL's on the previous timeline in the storage?

Reply:
I've added some comments to clarify this in 27a4884